### PR TITLE
New TokenPaymaster

### PR DESCRIPTION
**Pull request by forshtat**

### CODE REVIEW

File: contracts/samples/TokenPaymaster.sol
Comment (huaweigu) [inline contracts/samples/TokenPaymaster.sol:141]: minor, IIRC from our past debug experience, SafeERC20 uses banned [selfbalance opcode](https://github.com/OpenZeppelin/openzeppelin-contracts/blob/dfef6a68ee18dbd2e1f5a099061a3b8a0e404485/contracts/utils/Address.sol#L134)

File: contracts/samples/TokenPaymaster.sol
Comment (huaweigu) [inline contracts/samples/TokenPaymaster.sol]: I wonder if it's worth of attempting to transfer the owed token (preCharge - actualTokenNeeded) from sender here

File: contracts/samples/TokenPaymaster.sol
Comment (huaweigu) [inline contracts/samples/TokenPaymaster.sol]: Out of curiosity, how do we pick the upper bound (1900000)?

File: contracts/samples/TokenPaymaster.sol
Comment (huaweigu) [inline contracts/samples/TokenPaymaster.sol]: nitpick: maybe we shouldn't revert if preCharge == actualTokenNeeded (edge case)

File: contracts/samples/TokenPaymaster.sol
Comment (forshtat) [reply contracts/samples/TokenPaymaster.sol:141]: Thanks, I did not know that! Created a small PR in the OpenZeppelin repo, hope this is all there is to this issue:
https://github.com/OpenZeppelin/openzeppelin-contracts/pull/4271

I hope to avoid having to copy `SafeERC20.sol` library into the `account-abstraction` repo, but maybe we will have no choice.

File: contracts/samples/TokenPaymaster.sol
Comment (forshtat) [reply contracts/samples/TokenPaymaster.sol]: None of these hard-coded parameters are selected yet, but in my opinion the maximum possible value for the markup should just be there to prevent human error or malicious configuration.

For example, Pimlico Paymaster used `120e4` which sounds to me like it may be a bit too limiting:
https://github.com/pimlicolabs/erc20-paymaster-contracts/blob/c9911402a49928398441b07fb9caa9f4b4d0352f/src/PimlicoERC20Paymaster.sol#L68

File: contracts/samples/TokenPaymaster.sol
Comment (forshtat) [reply contracts/samples/TokenPaymaster.sol]: Right, thanks, will fix. To be clear, there is almost zero chance of that happening.

File: contracts/samples/TokenPaymaster.sol
Comment (forshtat) [reply contracts/samples/TokenPaymaster.sol]: I think this may actually work, will bring this up with the team. So far I did add an attempt to withdraw the owed amount (`actualTokenNeeded - preCharge`) from the UserOp sender to the Paymaster.

File: contracts/samples/TokenPaymaster.sol
Comment (livingrockrises) [reply contracts/samples/TokenPaymaster.sol]: maybe 2e6 is good upper limit

File: contracts/samples/TokenPaymaster.sol
Comment (livingrockrises) [inline contracts/samples/TokenPaymaster.sol:196]: Does this mean router swapping cost is dumped on to the users? 
since it is conditional postOp cost would change 
if not user who is paying for this computation - paymaster's gas deposit?

File: contracts/samples/TokenPaymaster.sol
Comment (forshtat) [reply contracts/samples/TokenPaymaster.sol]: @livingrockrises Agreed. Will set for 200% for now. Will also make it explicitly related to the `PRICE_DENOMINATOR` value.

File: contracts/samples/TokenPaymaster.sol
Comment (forshtat) [reply contracts/samples/TokenPaymaster.sol:196]: A single user should not get stuck with the bill for the Paymaster's swap operation on its own.
I am not sure but I think what we should do is to set the `POSTOP_COST` parameter such that it kind of spreads the cost of these periodic swaps over all other transactions.
Of course this way it may end up accumulating some loss or profit over time if left completely unattended.

File: contracts/samples/TokenPaymaster.sol
Comment (leekt) [reply contracts/samples/TokenPaymaster.sol:141]: maybe we can use solady in that case https://github.com/Vectorized/solady/blob/f9a6cc652dc8b37387590588a78644abfa4be3ac/src/utils/SafeTransferLib.sol#L238

File: contracts/samples/TokenPaymaster.sol
Comment (leekt) [inline contracts/samples/TokenPaymaster.sol:166]: This branch cannot be reached since there is a immediate return

File: contracts/samples/TokenPaymaster.sol
Comment (leekt) [inline contracts/samples/TokenPaymaster.sol:200]: STATICCALL to entrypoint is not allowed
```
Limitation on “CALL” opcodes (CALL, DELEGATECALL, CALLCODE, STATICCALL):
must not use value (except from account to the entrypoint)
must not revert with out-of-gas
destination address must have code (EXTCODESIZE>0)
cannot call EntryPoint’s methods, except depositFor (to avoid recursion)
```

File: contracts/samples/TokenPaymaster.sol
Comment (livingrockrises) [inline contracts/samples/TokenPaymaster.sol]: instead of using maxFeePerGas, does it make sense to use tx.gasprice ?

File: contracts/samples/TokenPaymaster.sol
Comment (livingrockrises) [reply contracts/samples/TokenPaymaster.sol:196]: If paymaster is always charging premium, POSTOP_COST can just be everything unaccounted on-chain on entry point and postOp - swap executionGas.

File: contracts/samples/TokenPaymaster.sol
Comment (livingrockrises) [inline contracts/samples/TokenPaymaster.sol:53]: what is the rationale behind using 1e26?

File: contracts/samples/TokenPaymaster.sol
Comment (forshtat) [reply contracts/samples/TokenPaymaster.sol]: You are right, thanks! However, we cannot just use `tx.gasprice` as the bundler can manipulate this value. I think we should use `userOp.gasPrice()` here. Updated.

File: contracts/samples/TokenPaymaster.sol
Comment (forshtat) [reply contracts/samples/TokenPaymaster.sol:166]: Thanks. Fixed.

File: contracts/samples/TokenPaymaster.sol
Comment (forshtat) [reply contracts/samples/TokenPaymaster.sol:53]: At this point constants are actually pretty random and will be finalised after we make some gas usage optimisations. Do you have suggestions on what value should be used and why?

File: contracts/samples/TokenPaymaster.sol
Comment (livingrockrises) [reply contracts/samples/TokenPaymaster.sol]: userop.gasPrice() touches BASEFEE which forbidden opcode. this would be fine in postOp(), but you shouldn't calculate it in validatePaymasterUserOp and pass it in context.
Instead pass maxPriorityFee and maxFee in context and calculate like below in the postOp

```
uint256 opGasPrice;
        unchecked {
            if (maxFeePerGas == maxPriorityFeePerGas) {
                opGasPrice = maxFeePerGas;
            } else {
                opGasPrice = Math.min(maxFeePerGas, maxPriorityFeePerGas + block.basefee);
            }
        }
```

File: contracts/samples/TokenPaymaster.sol
Comment (livingrockrises) [inline contracts/samples/TokenPaymaster.sol]: this will revert on the bundler
cc @leekt

File: contracts/samples/TokenPaymaster.sol
Comment (forshtat) [reply contracts/samples/TokenPaymaster.sol]: Right 😀

File: contracts/samples/TokenPaymaster.sol
Comment (leekt) [reply contracts/samples/TokenPaymaster.sol]: yup this will revert for sure

File: contracts/samples/TokenPaymaster.sol
Comment (leekt) [inline contracts/samples/TokenPaymaster.sol:185]: Not sure about this part. Won't this make `clientSuppliedPrice` invalid? `clientSuppliedPrice` should be the guard for how much token user is willing to pay but this part will end up forcing user to pay the full slippage

File: contracts/samples/TokenPaymaster.sol
Comment (drortirosh) [inline contracts/samples/TokenPaymaster.sol]: postop_cost depends on price of token transfer, which varies with different token implementation.
to be accurate, this is better be a constructor parameter

File: contracts/samples/TokenPaymaster.sol
Comment (drortirosh) [inline contracts/samples/TokenPaymaster.sol]: do we need setters? if the data is immutable, it the logic much cheaper.
There is no harm at re-deploy the paymaster with different configuration.

File: contracts/samples/TokenPaymaster.sol
Comment (drortirosh) [inline contracts/samples/TokenPaymaster.sol]: Better return validUntil as `cachedPriceTimestamp + maxCacheAge`

File: contracts/samples/TokenPaymaster.sol
Comment (drortirosh) [inline contracts/samples/TokenPaymaster.sol]: `abi.decode()` is that bad?
it surely much clearer to understand.

File: contracts/samples/utils/OracleHelper.sol
Comment (drortirosh) [inline contracts/samples/utils/OracleHelper.sol:81]: better be moved up, and avoid loading storage fields before the quick bail-out.

File: contracts/samples/TokenPaymaster.sol
Comment (forshtat) [reply contracts/samples/TokenPaymaster.sol]: This paymaster will accumulate client base by getting 'approval' in ERC-20 token used. Deploying a new paymaster just to switch to a different price oracle will lose all existing clients until they migrate their approval.
Do you want me to make the fields immutable and make the paymaster an upgradeable proxy instead?

File: contracts/samples/utils/OracleHelper.sol
Comment (forshtat) [reply contracts/samples/utils/OracleHelper.sol:81]: Fixed.

File: contracts/samples/TokenPaymaster.sol
Comment (forshtat) [reply contracts/samples/TokenPaymaster.sol]: Fixed. I was trying to same some bytes by doing `abi.encodePacked`.

File: contracts/samples/TokenPaymaster.sol
Comment (forshtat) [reply contracts/samples/TokenPaymaster.sol]: Done.

File: contracts/samples/TokenPaymaster.sol
Comment (forshtat) [reply contracts/samples/TokenPaymaster.sol]: @drortirosh
Bumping this one. Should we?

File: contracts/samples/TokenPaymaster.sol
Comment (forshtat) [reply contracts/samples/TokenPaymaster.sol]: Done.

File: contracts/samples/TokenPaymaster.sol
Comment (forshtat) [reply contracts/samples/TokenPaymaster.sol:185]: Well, yes, with this implementation the user-supplied price can only increase the pre-charge for the user, but does not set an upper limit of the token price, so once the users approve this paymaster they basically agree to pay for gas at any current market price. 

@drortirosh what do you think about this feature?

CC: @huaweigu

File: contracts/samples/TokenPaymaster.sol
Comment (drortirosh) [inline contracts/samples/TokenPaymaster.sol:145]: storage optimization: instead referencing 2 storage slots on each call, use 
`cachedPriceExpirationTimestamp = timestamp + priceMaxAge`

File: test/samples/OracleHelper.test.ts
Comment (drortirosh) [inline test/samples/OracleHelper.test.ts:110]: ^^ remove "only" from `describe.only('OracleHelper')`

File: contracts/index.ts
Comment (drortirosh) [inline contracts/index.ts:1]: true, should be removed (along with the code to use it during "yarn publish"), but unrelated to TokenPaymaster.

File: contracts/test/TestUniswap.sol
Comment (drortirosh) [inline contracts/test/TestUniswap.sol:48]: unused?

File: contracts/samples/TokenPaymaster.sol
Comment (drortirosh) [inline contracts/samples/TokenPaymaster.sol]: I think this is `_cachedPrice`, not `cachedPrice`

File: contracts/samples/TokenPaymaster.sol
Comment (drortirosh) [reply contracts/samples/TokenPaymaster.sol:196]: @livingrockrises 
A paymaster is expected to set `refundPostopCost` to actual cost of postOp, with faction of the cost of the uniswap, since the swap is expected to happen only every "N" transaction.

File: contracts/samples/TokenPaymaster.sol
Comment (drortirosh) [reply contracts/samples/TokenPaymaster.sol]: As a user, I trust the config values the paymaster have - but I really can't fully trust it, since it can change its configuration at will  (e.g. increase the postOp cost, or switch to a malicious oracle)

I want to trust an address given by some authority, and know this address can't change.
Better yet - I want to know this address define all parameters values.

Also other parameters (like minEPbalance) are not expected to be modified much.

**Review by drortirosh — COMMENTED** [review]
Should also add gas-check test in `gas-calc/`

File: contracts/index.ts
Comment (forshtat) [reply contracts/index.ts:1]: This was messing up the environment for me though.

File: contracts/test/TestUniswap.sol
Comment (forshtat) [reply contracts/test/TestUniswap.sol:48]: Why not? This is used by the `UniswapHelper.sol` file:

```
    function unwrapWeth(uint256 amount) internal {
        IPeripheryPayments(address(uniswap)).unwrapWETH9(amount, address(this));
    }
```

File: contracts/samples/TokenPaymaster.sol
Comment (forshtat) [reply contracts/samples/TokenPaymaster.sol:145]: Please elaborate where does the `timestamp` variable come from?

File: contracts/samples/utils/OracleHelper.sol
Comment (drortirosh) [inline contracts/samples/utils/OracleHelper.sol:70]: leaving this method (called only from constructor) means that we still pay the SLOAD when using these parameters.
(and you can't make a member "immutable" if it is set from a private method)

Currently, the paymaster uses 15 slots (14, if we exclude "owner")
(checked with `solc --storage-layout`)
So it uses 14 storage slots on each operation, which is at minimum 14*2100 = 29400 gas .
But I suggest adding a "gascalc" test before optimizing it.

**Review by drortirosh — APPROVED** [review]

File: contracts/samples/TokenPaymaster.sol
Comment (livingrockrises) [reply contracts/samples/TokenPaymaster.sol]: at Biconomy we are returning a fee quote to the sender and allow them to update userop.calldata with the approval transaction with necessary amount each time. then partialUserOp is filled with paymasterAndData and signature. SDK account APIs can facilitate this for decoding calldata and append approval transaction in a batch and have new callData, callGasLimit

**Review by Blackjack789 — APPROVED** [review]

**Pull request merged**
